File: apps/desktop-linux/src/export_page.rs
use gtk4 as gtk;
use libadwaita as adw;
use gtk::gdk;
use gtk::glib;
use adw::prelude::*;
use std::rc::Rc;
use std::cell::RefCell;

use crate::types::AppState;
use crate::exporter;
use crate::ui_helpers::{apply_widget_css, set_margin_all};

pub fn build_export_page(
    state: Rc<RefCell<AppState>>,
    refresh_export_view: Rc<RefCell<Option<Rc<dyn Fn()>>>>,
) -> gtk::Widget {
    let main_box = gtk::Box::new(gtk::Orientation::Vertical, 12);
    set_margin_all(&main_box, 12);

    // Header Format Selector
    let format_header = gtk::Box::new(gtk::Orientation::Horizontal, 8);
    format_header.set_halign(gtk::Align::Center);
    main_box.append(&format_header);

    let current_format = Rc::new(RefCell::new("css".to_string()));

    let text_view = gtk::TextView::builder()
        .monospace(true)
        .editable(false)
        .cursor_visible(false)
        .wrap_mode(gtk::WrapMode::WordChar)
        .build();
    text_view.add_css_class("mono-text");
    apply_widget_css(&text_view, "* { padding: 12px; background-color: @theme_bg_color; border-radius: 8px; border: 1px solid rgba(0,0,0,0.1); }");

    let scrolled = gtk::ScrolledWindow::builder()
        .vexpand(true)
        .child(&text_view)
        .build();
    main_box.append(&scrolled);

    // Actions Footer
    let actions_box = gtk::Box::new(gtk::Orientation::Horizontal, 12);
    actions_box.set_margin_top(8);
    let btn_copy = gtk::Button::builder()
        .label("Copy to Clipboard")
        .hexpand(true)
        .build();
    btn_copy.add_css_class("suggested-action");
    let btn_save = gtk::Button::builder()
        .label("Save as File…")
        .hexpand(true)
        .build();
    actions_box.append(&btn_copy);
    actions_box.append(&btn_save);
    main_box.append(&actions_box);

    let formats = [
        ("css", "CSS vars"),
        ("scss", "SCSS vars"),
        ("json", "Style Dict"),
        ("tailwind", "Tailwind"),
        ("android", "Android XML"),
    ];

    // Redraw logic
    let refresh_all = {
        let state = state.clone();
        let current_format = current_format.clone();
        let text_view = text_view.clone();
        let main_box = main_box.clone();

        move || {
            let s = state.borrow();
            let is_empty = s.token_groups.is_empty() || s.token_groups.iter().all(|g| g.tokens.is_empty());

            if is_empty {
                text_view.buffer().set_text("Add tokens in the Tokens tab to generate exports.");
                main_box.set_sensitive(false);
            } else {
                main_box.set_sensitive(true);
                let format = current_format.borrow().clone();
                let output = match format.as_str() {
                    "css" => exporter::export_css(&s.token_groups, &s.palettes),
                    "scss" => exporter::export_scss(&s.token_groups, &s.palettes),
                    "json" => exporter::export_json(&s.token_groups, &s.palettes),
                    "tailwind" => exporter::export_tailwind(&s.token_groups, &s.palettes),
                    "android" => exporter::export_android_xml(&s.token_groups, &s.palettes),
                    _ => String::new(),
                };
                text_view.buffer().set_text(&output);
            }
        }
    };

    // Format buttons connections
    for (fmt_id, label) in formats {
        let btn = gtk::Button::builder()
            .label(label)
            .build();
        format_header.append(&btn);

        let fmt_id_str = fmt_id.to_string();
        let current_format_clone = current_format.clone();
        let refresh_clone = refresh_all.clone();
        btn.connect_clicked(move |_| {
            *current_format_clone.borrow_mut() = fmt_id_str.clone();
            refresh_clone();
        });
    }

    // Copy Connection
    {
        let text_view = text_view.clone();
        btn_copy.connect_clicked(move |btn| {
            let buffer = text_view.buffer();
            let (start, end) = buffer.bounds();
            let text = buffer.text(&start, &end, false).to_string();

            let clipboard = gdk::Display::default().unwrap().clipboard();
            clipboard.set_text(&text);

            let old_label = btn.label().unwrap_or_default();
            btn.set_label("Copied!");
            let btn_clone = btn.clone();
            glib::timeout_add_local(std::time::Duration::from_secs(2), move || {
                btn_clone.set_label(&old_label);
                glib::ControlFlow::Break
            });
        });
    }

    // Save Connection
    {
        let text_view = text_view.clone();
        let current_format = current_format.clone();
        btn_save.connect_clicked(move |_| {
            let buffer = text_view.buffer();
            let (start, end) = buffer.bounds();
            let text = buffer.text(&start, &end, false).to_string();

            let ext = current_format.borrow().clone();
            let file_name = format!("tokens.{}", if ext == "android" { "xml" } else if ext == "tailwind" { "ts" } else { &ext });

            let dialog = gtk::FileDialog::builder()
                .title("Save Exported Tokens")
                .initial_name(&file_name)
                .build();

            dialog.save(None::<&gtk::Window>, gtk::gio::Cancellable::NONE, move |res| {
                if let Ok(file) = res {
                    if let Some(path) = file.path() {
                        let _ = std::fs::write(path, &text);
                    }
                }
            });
        });
    }

    *refresh_export_view.borrow_mut() = Some(Rc::new(refresh_all));

    main_box.upcast::<gtk::Widget>()
}

File: apps/desktop-linux/src/export_page.cpp
#include "export_page.h"
#include "exporter.h"
#include "ui_helpers.h"
#include <fstream>
#include <iostream>

GtkWidget* build_export_page(
    std::shared_ptr<AppState> state,
    std::shared_ptr<std::function<void()>> refresh_export_view
) {
    GtkWidget* main_box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 12);
    set_margin_all(main_box, 12);

    // Header Format Selector
    GtkWidget* format_header = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 8);
    gtk_widget_set_halign(format_header, GTK_ALIGN_CENTER);
    gtk_box_append(GTK_BOX(main_box), format_header);

    auto current_format = std::make_shared<std::string>("css");

    GtkWidget* text_view = gtk_text_view_new();
    gtk_text_view_set_monospace(GTK_TEXT_VIEW(text_view), TRUE);
    gtk_text_view_set_editable(GTK_TEXT_VIEW(text_view), FALSE);
    gtk_text_view_set_cursor_visible(GTK_TEXT_VIEW(text_view), FALSE);
    gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW(text_view), GTK_WRAP_WORD_CHAR);
    gtk_widget_add_css_class(text_view, "mono-text");
    apply_widget_css(text_view, "* { padding: 12px; background-color: @theme_bg_color; border-radius: 8px; border: 1px solid rgba(0,0,0,0.1); }");

    GtkWidget* scrolled = gtk_scrolled_window_new();
    gtk_widget_set_vexpand(scrolled, TRUE);
    gtk_scrolled_window_set_child(GTK_SCROLLED_WINDOW(scrolled), text_view);
    gtk_box_append(GTK_BOX(main_box), scrolled);

    // Actions Footer
    GtkWidget* actions_box = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 12);
    gtk_widget_set_margin_top(actions_box, 8);

    GtkWidget* btn_copy = gtk_button_new_with_label("Copy to Clipboard");
    gtk_widget_set_hexpand(btn_copy, TRUE);
    gtk_widget_add_css_class(btn_copy, "suggested-action");

    GtkWidget* btn_save = gtk_button_new_with_label("Save as File…");
    gtk_widget_set_hexpand(btn_save, TRUE);

    gtk_box_append(GTK_BOX(actions_box), btn_copy);
    gtk_box_append(GTK_BOX(actions_box), btn_save);
    gtk_box_append(GTK_BOX(main_box), actions_box);

    struct FormatInfo {
        const char* id;
        const char* label;
    };
    FormatInfo formats[] = {
        {"css", "CSS vars"},
        {"scss", "SCSS vars"},
        {"json", "Style Dict"},
        {"tailwind", "Tailwind"},
        {"android", "Android XML"}
    };

    // Redraw logic
    auto refresh_all = [state, current_format, text_view, main_box]() {
        bool is_empty = state->token_groups.empty();
        if (!is_empty) {
            is_empty = true;
            for (const auto& g : state->token_groups) {
                if (!g.tokens.empty()) {
                    is_empty = false;
                    break;
                }
            }
        }

        GtkTextBuffer* buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(text_view));
        if (is_empty) {
            gtk_text_buffer_set_text(buffer, "Add tokens in the Tokens tab to generate exports.", -1);
            gtk_widget_set_sensitive(main_box, FALSE);
        } else {
            gtk_widget_set_sensitive(main_box, TRUE);
            std::string format = *current_format;
            std::string output;
            if (format == "css") {
                output = export_css(state->token_groups, state->palettes);
            } else if (format == "scss") {
                output = export_scss(state->token_groups, state->palettes);
            } else if (format == "json") {
                output = export_json(state->token_groups, state->palettes);
            } else if (format == "tailwind") {
                output = export_tailwind(state->token_groups, state->palettes);
            } else if (format == "android") {
                output = export_android_xml(state->token_groups, state->palettes);
            }
            gtk_text_buffer_set_text(buffer, output.c_str(), -1);
        }
    };

    // Format buttons connections
    for (const auto& fmt : formats) {
        GtkWidget* btn = gtk_button_new_with_label(fmt.label);
        gtk_box_append(GTK_BOX(format_header), btn);

        connect_clicked(btn, [current_format, fmt_id = std::string(fmt.id), refresh_all](GtkButton* button) {
            *current_format = fmt_id;
            refresh_all();
        });
    }

    // Copy Connection
    connect_clicked(btn_copy, [text_view](GtkButton* btn) {
        GtkTextBuffer* buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(text_view));
        GtkTextIter start, end;
        gtk_text_buffer_get_bounds(buffer, &start, &end);
        char* text = gtk_text_buffer_get_text(buffer, &start, &end, FALSE);

        GdkClipboard* clipboard = gdk_display_get_clipboard(gdk_display_get_default());
        gdk_clipboard_set_text(clipboard, text);
        g_free(text);

        gtk_button_set_label(btn, "Copied!");
        
        struct TimeoutData {
            GtkButton* btn;
        };
        auto* tdata = new TimeoutData{btn};
        g_timeout_add_seconds(2, [](gpointer data) -> gboolean {
            auto* td = static_cast<TimeoutData*>(data);
            gtk_button_set_label(td->btn, "Copy to Clipboard");
            delete td;
            return G_SOURCE_REMOVE;
        }, tdata);
    });

    // Save Connection
    connect_clicked(btn_save, [text_view, current_format](GtkButton* btn) {
        GtkTextBuffer* buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(text_view));
        GtkTextIter start, end;
        gtk_text_buffer_get_bounds(buffer, &start, &end);
        char* text = gtk_text_buffer_get_text(buffer, &start, &end, FALSE);
        std::string text_str = text;
        g_free(text);

        std::string ext = *current_format;
        std::string ext_val = ext;
        if (ext == "android") ext_val = "xml";
        else if (ext == "tailwind") ext_val = "ts";
        std::string file_name = "tokens." + ext_val;

        GtkFileDialog* dialog = gtk_file_dialog_new();
        gtk_file_dialog_set_title(dialog, "Save Exported Tokens");
        gtk_file_dialog_set_initial_name(dialog, file_name.c_str());

        struct SaveData {
            std::string text;
        };
        auto* sdata = new SaveData{text_str};

        gtk_file_dialog_save(dialog, NULL, NULL, [](GObject* source_object, GAsyncResult* res, gpointer data) {
            GtkFileDialog* file_dialog = GTK_FILE_DIALOG(source_object);
            auto* sd = static_cast<SaveData*>(data);
            GError* error = NULL;
            GFile* file = gtk_file_dialog_save_finish(file_dialog, res, &error);
            if (file) {
                char* path = g_file_get_path(file);
                if (path) {
                    std::ofstream ofs(path);
                    if (ofs.is_open()) {
                        ofs << sd->text;
                    }
                    g_free(path);
                }
                g_object_unref(file);
            }
            delete sd;
        }, sdata);
    });

    *refresh_export_view = refresh_all;

    return main_box;
}
